File: web/src/parse.ts
import parseBlockInlines from "./parse/parseBlockInlines";
import { parseLine } from "./parse/parseLine";
import type BlockParserState from "./types/BlockParserState";
import type MarkdownNode from "./types/MarkdownNode";
import type RuleSet from "./types/RuleSet";
import { DASH_CODE, NEW_LINE_CODE } from "./utils/charCodes";
import isNewLine from "./utils/isNewLine";
import isSpace from "./utils/isSpace";
import newBlock from "./utils/newBlock";

export default function parse(src: string, rules: RuleSet): MarkdownNode {
	let document = newBlock("document", 0, 1, "", 0);

	// Skip empty lines at the start
	let i = 0;
	let index = 0;
	for (; index < src.length; index++) {
		let nextCharCode = src.charCodeAt(index);
		if (!isSpace(nextCharCode)) {
			break;
		} else if (isNewLine(nextCharCode)) {
			i = index + 1;
		}
	}

	// Process frontmatter if found
	let frontmatter: string | undefined;
	if (src.charCodeAt(index) === DASH_CODE) {
		frontmatter = extractFrontMatter(document, src, index);
		if (frontmatter !== undefined) {
			i = index + frontmatter.length;
		}
	}

	let state: BlockParserState = {
		rules: rules.blocks,
		rulesMap: new Map(rules.blocks.map((b) => [b.name, b])),
		src,
		i,
		line: 0,
		lineStart: 0,
		indent: 0,
		isEscaped: false,
		maybeContinue: false,
		hasBlankLine: false,
		openNodes: [document],
		refs: {},
		footnotes: {},
	};

	// Stage 1 -- parse each line into blocks
	while (state.i < state.src.length) {
		parseLine(state);
	}

	// Close the remaining open nodes
	let j = state.openNodes.length;
	while (j--) {
		let openNode = state.openNodes[j];
		openNode.length = state.i - openNode.index;
		let rule = state.rulesMap.get(openNode.type);
		if (rule?.closeNode !== undefined) {
			rule.closeNode(state, openNode);
		}
	}

	// Stage 2 -- parse the inlines for each block
	parseBlockInlines(document, rules.inlines, state.refs, state.footnotes);

	if (frontmatter !== undefined) {
		document.info = frontmatter;
	}

	return document;
}

function extractFrontMatter(document: MarkdownNode, src: string, index: number) {
	let frontmatter: string | undefined;

	if (src.charCodeAt(index) === DASH_CODE && /^---\s*\r?\n/.test(src.substring(index))) {
		let contentEnd = -1;
		for (let j = index + 3; j < src.length; j++) {
			if (src.charCodeAt(j) === DASH_CODE && /^---\s*\r?\n/.test(src.substring(j))) {
				contentEnd = src.length;
				for (let k = j + 3; k < src.length; k++) {
					if (src.charCodeAt(k) === NEW_LINE_CODE) {
						contentEnd = k;
						break;
					}
				}
			}
		}
		if (contentEnd !== -1) {
			frontmatter = src.substring(index, contentEnd);
			let i = contentEnd;
			document.line = src.substring(0, i).split("\n").length;
		}
	}

	return frontmatter;
}

File: zig/src/parse/parse.zig
const std = @import("std");
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const BlockParserState = @import("../types/BlockParserState.zig").BlockParserState;
const RuleSet = @import("../types/RuleSet.zig").RuleSet;
const BlockRule = @import("../types/BlockRule.zig").BlockRule;
const isSpace = @import("../utils/isSpace.zig").isSpace;
const isNewLine = @import("../utils/isNewLine.zig").isNewLine;
const newBlock = @import("../utils/newBlock.zig").newBlock;
const newInline = @import("../utils/newInline.zig").newInline;
const parseLine = @import("./parseLine.zig").parseLine;
const parseBlockInlines = @import("./parseBlockInlines.zig").parseBlockInlines;
const extractFrontMatter = @import("../utils/extractFrontMatter.zig").extractFrontMatter;

pub fn parse(allocator: std.mem.Allocator, src: []const u8, rules: RuleSet) !*MarkdownNode {
    const document = newBlock(allocator, "document", 0, 1, "", 0) catch unreachable;

    var i: usize = 0;
    var index: usize = 0;
    while (index < src.len) : (index += 1) {
        if (!isSpace(src[index])) {
            break;
        } else if (isNewLine(src[index])) {
            i = index + 1;
        }
    }

    var frontmatter: ?[]const u8 = null;
    if (i < src.len and src[i] == '-') {
        frontmatter = extractFrontMatter(allocator, document, src, index) catch null;
        if (frontmatter) |fm| {
            i = index + fm.len;
        }
    }

    var openNodes = std.ArrayList(*MarkdownNode).initCapacity(allocator, 1) catch unreachable;
    openNodes.append(allocator, document) catch unreachable;

    var rulesMap = std.StringHashMap(*const BlockRule).init(allocator);
    for (rules.blocks) |rule| {
        try rulesMap.put(rule.name, rule);
    }

    var state = BlockParserState{
        .allocator = allocator,
        .rules = rules.blocks,
        .rulesMap = rulesMap,
        .src = src,
        .i = i,
        .line = 1,
        .lineStart = 0,
        .indent = 0,
        .isEscaped = false,
        .maybeContinue = false,
        .hasBlankLine = false,
        .openNodes = openNodes,
        .refs = std.StringHashMap(@import("../types/LinkReference.zig").LinkReference).init(allocator),
        .footnotes = std.StringHashMap(@import("../types/FootnoteReference.zig").FootnoteReference).init(allocator),
    };
    defer state.openNodes.deinit(allocator);
    defer {
        var iter = state.refs.iterator();
        while (iter.next()) |entry| {
            state.allocator.free(entry.key_ptr.*);
            state.allocator.free(entry.value_ptr.url);
            state.allocator.free(entry.value_ptr.title);
        }
        state.refs.deinit();
    }
    defer {
        var iter = state.footnotes.iterator();
        while (iter.next()) |entry| {
            // Free the label (same as key, but stored separately in FootnoteReference)
            state.allocator.free(entry.value_ptr.label);
        }
        state.footnotes.deinit();
    }
    defer {
        var iter = state.rulesMap.iterator();
        while (iter.next()) |_| {
            // No need to free the key - it's a string literal
        }
        state.rulesMap.deinit();
    }

    while (state.i < state.src.len) {
        const start_i = state.i;
        parseLine(&state);
        // Debug
        // Safety check: ensure we made progress
        if (state.i == start_i) {
            state.i += 1; // Force progress to avoid infinite loop
        }
    }

    var j = state.openNodes.items.len;
    while (j > 0) : (j -= 1) {
        const openNode = state.openNodes.items[j - 1];
        openNode.length = state.i - openNode.index;
        if (state.rulesMap.get(openNode.type)) |rule| {
            if (rule.closeNode) |closeFn| {
                closeFn(&state, openNode);
            }
        }
    }

    parseBlockInlines(allocator, document, rules.inlines, state.refs, state.footnotes) catch |err| {
        std.debug.print("Error parsing block inlines: {s}\n", .{@errorName(err)});
    };

    if (frontmatter) |fm| {
        document.info = fm;
    }

    return document;
}
